src/Instrumentation/Laravel/src/Hooks/Illuminate/Contracts/Queue/Queue.php

```diff
@@ -0,0 +1,145 @@
+declare(strict_types=1);
+
+namespace OpenTelemetry\Contrib\Instrumentation\Laravel\Hooks\Illuminate\Contracts\Queue;
+
+use DateInterval;
+use DateTime;
+use DateTimeInterface;
+use Illuminate\Contracts\Queue\Queue as QueueContract;
+use OpenTelemetry\API\Trace\SpanKind;
+use OpenTelemetry\Context\Context;
+use OpenTelemetry\Contrib\Instrumentation\Laravel\Hooks\HookInstance;
+use OpenTelemetry\Contrib\Instrumentation\Laravel\Hooks\Illuminate\Queue\AttributesBuilder;
+use OpenTelemetry\Contrib\Instrumentation\Laravel\Hooks\PostHookHandler;
+use function OpenTelemetry\Instrumentation\hook;
+use OpenTelemetry\SDK\Common\Time\ClockFactory;
+use OpenTelemetry\SDK\Common\Time\ClockInterface;
+use OpenTelemetry\SemConv\TraceAttributes;
+use OpenTelemetry\SemConv\TraceAttributeValues;
+use Throwable;
+
+class Queue
+{
+    use AttributesBuilder;
+    use HookInstance;
+    use PostHookHandler;
+
+    public function instrument(): void
+    {
+        $this->hookBulk();
+        $this->hookLater();
+        $this->hookPushRaw();
+    }
+
+    protected function hookBulk(): bool
+    {
+        return hook(
+            QueueContract::class,
+            'bulk',
+            pre: function (QueueContract $queue, array $params, string $class, string $function, ?string $filename, ?int $lineno) {
+                $attributes = array_merge([
+                    TraceAttributes::CODE_FUNCTION => $function,
+                    TraceAttributes::CODE_NAMESPACE => $class,
+                    TraceAttributes::CODE_FILEPATH => $filename,
+                    TraceAttributes::CODE_LINENO => $lineno,
+                    TraceAttributes::MESSAGING_BATCH_MESSAGE_COUNT => count($params[0] ?? []),
+                ], $this->contextualMessageSystemAttributes($queue, []));
+
+                /** @psalm-suppress ArgumentTypeCoercion */
+                $span = $this->instrumentation
+                    ->tracer()
+                    ->spanBuilder(vsprintf('%s %s', [
+                        method_exists($queue, 'getQueue') ? $queue->getQueue($params[2] ?? null) : $queue->getConnectionName(),
+                        TraceAttributeValues::MESSAGING_OPERATION_PUBLISH,
+                    ]))
+                    ->setSpanKind(SpanKind::KIND_PRODUCER)
+                    ->setAttributes($attributes)
+                    ->startSpan();
+
+                Context::storage()->attach($span->storeInContext(Context::getCurrent()));
+
+                return $params;
+            },
+            post: function (QueueContract $queue, array $params, $returnValue, ?Throwable $exception) {
+                $this->endSpan($exception);
+            },
+        );
+    }
+
+    protected function hookLater(): bool
+    {
+        return hook(
+            QueueContract::class,
+            'later',
+            pre: function (QueueContract $queue, array $params, string $class, string $function, ?string $filename, ?int $lineno) {
+                $clock = ClockFactory::getDefault();
```

Comment:
Is this permissible usage of an SDK utility?

Reply:
I think it falls foul of the "should not depend on the SDK" part of spec, ie this package should work without the SDK installed.
I created https://github.com/open-telemetry/opentelemetry-php/pull/1281 to migrate clock into API, but note the downside that using it would mean dropping 8.0 support (unless it's important enough to try to back-patch it)

Reply:
https://opentelemetry.io/docs/concepts/instrumentation/libraries/#opentelemetry-api is why the library shouldn't directly use any part of the SDK.

> Libraries should only use the OpenTelemetry API

Reply:
I've removed the usage of the clock component for now. I simplified the test and hope to revisit this once the clock components are available in the API! :+1: